test/handler/async.test.mo

```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
+assert handler.journalLength() == inc(1); // #debited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase deposit again
+await ledger.set_balance(15);
+assert (await* handler.notify(user1)) == ?(15, 5); // deposit = 15, credit = 5
+assert_state(15, 0, 1);
+assert handler.journalLength() == inc(2); // #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while notify is underway (and item still in queue)
+// scenario 2: old_fee < previous <= new_fee < latest
+await ledger.set_balance(20);
+await ledger.lock_balance();
+let f2 = async { await* handler.notify(user1) }; // would return ?(5,10) at old fee
+await ledger.set_fee(15);
+ignore await* handler.updateFee();
+assert handler.journalLength() == inc(1); // #feeUpdated, not #debited because user1 is locked
+assert_state(15, 0, 1); // state still unchanged
+await ledger.release_balance(); // let notify return
+assert (await f2) == ?(20, 5); // credit = latest - new_fee
+assert_state(20, 0, 1); // state should have changed
+assert handler.journalLength() == inc(3); // #debited (recalculation), #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// call multiple notify() simultaneously
+await ledger.lock_balance();
+let arr = [async { await* handler.notify(user1) }, async { await* handler.notify(user1) }, async { await* handler.notify(user1) }];
+await ledger.release_balance(); // let notify return
+assert (await arr[0]) == ?(0, 5); // first notify() should return state
+assert (await arr[1]) == null; // should return null
+assert (await arr[2]) == null; // should return null
+assert_state(20, 0, 1); // state unchanged because deposit has not changed
+assert handler.journalLength() == inc(0);
+print("tree lookups = " # debug_show handler.lookups());
+
+// only 1 consolidation process can be triggered for same user at same time
+var transfer_count = await ledger.transfer_count();
+let f3 = async { await* handler.trigger() };
+let f4 = async { await* handler.trigger() };
+await f3;
+await f4;
+assert ((await ledger.transfer_count())) == transfer_count + 1; // only 1 transfer call has been made
+assert_state(0, 5, 0); // consolidation successful
+assert handler.journalLength() == inc(1); // #consolidated
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while deposit is being consolidated
+// scenario 1: old_fee < deposit < new_fee
+// consolidation should fail and deposit should be reset
+await ledger.set_balance(20);
+assert (await* handler.notify(user1)) == ?(20, 10); // deposit = 20, credit = 10
+assert handler.journalLength() == inc(2); // #credited, #newDeposit
+assert_state(20, 5, 1);
+await ledger.lock_transfer();
+let f5 = async { await* handler.trigger() };
+await ledger.set_fee(25);
+await ledger.set_response(#Err(#BadFee { expected_fee = 25 }));
+await ledger.release_transfer(); // let transfer return
+await f5;
+assert_state(0, 5, 0); // consolidation failed
```

Comment:
We should make one more assertion here, that the user's credit has been reduced (debited).